language/move-ir/types/src/ast.rs

```diff
@@ -164,35 +170,52 @@ pub enum Kind {
 //**************************************************************************************************
 
 /// The type of a single value
-#[derive(Debug, PartialEq, Clone)]
+#[derive(Debug, Display, PartialEq, Clone)]
 pub enum Type {
     /// `address`
+    #[display(fmt = "address")]
     Address,
     /// `u8`
+    #[display(fmt = "u8")]
     U8,
     /// `u64`
+    #[display(fmt = "u64")]
     U64,
     /// `u128`
+    #[display(fmt = "u128")]
     U128,
     /// `bool`
+    #[display(fmt = "bool")]
     Bool,
     /// `bytearray`
+    #[display(fmt = "bytearray")]
     ByteArray,
     /// A module defined struct
-    Struct(QualifiedStructIdent, Vec<Type>),
+    #[display(fmt = "{}{}", _0, _1)]
+    Struct(QualifiedStructIdent, TypeActuals),
     /// A reference type, the bool flag indicates whether the reference is mutable
+    #[display(fmt = "&{}{}", "if *_0 { \"mut \" } else { \"\" }", _1)]
     Reference(bool, Box<Type>),
     /// A type parameter
+    #[display(fmt = "{}", _0)]
     TypeParameter(TypeVar_),
 }
 
+/// A list of actual types.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct TypeActuals {
```

Comment:
What is the reason to add this separate struct? It doesn't seem like it is anything more than just a vector of types. I suppose you might get a marginal improvement in type safety, but it's at the expense of transparency and simplicity. Is there some other motivation that I'm missing?

Reply:
Didn't see this (thought it was just a soft type alias based on the usage). I'm going to have to agree with Bob that I don't think this change buys much

Reply:
In other words, I think a soft type alias would do much better here

Reply:
I disagree -- a struct is much better here. From the commit message:

> This has two purposes:

> 1. Makes it easy to distinguish between type actuals and the Vec<Type> used as the return type from functions.
> 2. Allows for a cleaner `Display` impl.

Reply:
Ah, I had missed the explanation in the commit message.... I don't know much about Rust's Display thingy, so I'll defer to the rest of you on that. My personal inclination would be to keep it simple.

Reply:
I think (1) is also achieved by a soft type alias. And often times in the compiler you actually WANT the two to be interchangeable (i.e. you want to easily have code that works for both without jumping through a trait)

I agree with (2) that the implementation is cleaner, but I don't think the benefit outweighs the cost

Reply:
> (i.e. you want to easily have code that works for both without jumping through a trait)

That trait is `IntoIterator`, and there was just one place where it was necessary to do that. It was quite easy to do in any case.

> I agree with (2) that the implementation is cleaner, but I don't think the benefit outweighs the cost

I do...

Reply:
>> I agree with (2) that the implementation is cleaner, but I don't think the benefit outweighs the cost

> I do...

I'm not really clear on the benefit here other than being able to declare additional impls easily against this type. 

Often the compiler, you want less constructs in your AST, not more. The reason being is you want as writing passes you want similar constructs to have the same code, as it forces you into a space of being more principled with your judgements/rules/translations, and forces you to think if you really need an ad hoc pass. In this case, I want the type actuals to be the same as any other vector of types, that is an intentional and desirable design choice. The reason being is that any vector/list of types should behave in the same way in their principles regardless of the usage. This change would make the default/easy option the ad hoc one, and would force you out of the default world of shared, principled passes. 

In short, this design choice to share `Vec<Types>` everywhere might seem lazy or unsafe in other software applications, but it is a specific and intentional choice for the compiler.